pkg/cloud/gcp/actuators/machine/reconciler.go

```diff
@@ -117,30 +114,63 @@ func (r *Reconciler) create() error {
+// Create creates a new cloud instance
+func (r *Reconciler) create() error {
+	if err := validateMachine(*r.machine, *r.providerSpec); err != nil {
+		return fmt.Errorf("failed validating machine provider spec: %v", err)
+	}
+
+	if r.machine.Annotations == nil {
+		r.machine.Annotations = map[string]string{}
+	}
+
+	instance, err := r.instanceGet()
 	if err != nil {
-		if reconcileWithCloudError := r.reconcileMachineWithCloudState(&v1beta1.GCPMachineProviderCondition{
-			Type:    v1beta1.MachineCreated,
-			Reason:  machineCreationFailedReason,
-			Message: err.Error(),
-			Status:  corev1.ConditionFalse,
-		}); reconcileWithCloudError != nil {
-			klog.Errorf("Failed to reconcile machine with cloud state: %v", reconcileWithCloudError)
-		}
-		return fmt.Errorf("failed to create instance via compute service: %v", err)
+		return err
 	}
-	if op, err := r.waitUntilOperationCompleted(zone, operation.Name); err != nil {
-		if reconcileWithCloudError := r.reconcileMachineWithCloudState(&v1beta1.GCPMachineProviderCondition{
-			Type:    v1beta1.MachineCreated,
-			Reason:  machineCreationFailedReason,
-			Message: err.Error(),
-			Status:  corev1.ConditionFalse,
-		}); reconcileWithCloudError != nil {
-			klog.Errorf("Failed to reconcile machine with cloud state: %v", reconcileWithCloudError)
+
+	var operation *compute.Operation
+
+	if instance != nil && haveCreateOperationInProgress(r.machine) {
+		operation, err = r.computeService.ZoneOperationsGet(r.projectID, r.providerSpec.Zone, r.machine.Annotations[pendingCreateKey])
+		if err != nil {
+			return err
 		}
-		return fmt.Errorf("failed to wait for create operation via compute service. Operation status: %v. Error: %v", op, err)
+	} else {
+		instance, err = r.createInstanceFromMachineScope()
+		if err != nil {
+			return err
+		}
+		operation, err = r.computeService.InstancesInsert(r.projectID, r.providerSpec.Zone, instance)
+		if err != nil {
+			if reconcileWithCloudError := r.reconcileMachineWithCloudState(&v1beta1.GCPMachineProviderCondition{
+				Type:    v1beta1.MachineCreated,
+				Reason:  machineCreationFailedReason,
+				Message: err.Error(),
+				Status:  corev1.ConditionFalse,
+			}); reconcileWithCloudError != nil {
+				klog.Errorf("Failed to reconcile machine with cloud state: %v", reconcileWithCloudError)
+			}
+			return fmt.Errorf("failed to create instance via compute service: %v", err)
+		}
+		r.machine.Annotations[pendingCreateKey] = fmt.Sprintf("%v", operation.Id)
 	}
-	return r.reconcileMachineWithCloudState(nil)
+
+	klog.Infof("Instance create operation #%v status=%q for machine %q", operation.Id, operation.Status, r.machine.Name)
+	if err := r.reconcileMachineWithCloudState(nil); err != nil {
+		return err
+	}
+
+	if operation.Status != "DONE" {
+		klog.Infof("Instance create operation #%v incomplete, returning an error to requeue", operation.Id)
+		return &clustererror.RequeueAfterError{RequeueAfter: requeuePeriod}
+	}
+
+	delete(r.machine.Annotations, pendingCreateKey)
```

Comment:
We should add a comment here that (hopefully) scope.close() will sync object state to make this remove effective.

Reply:
The whole `scope.Close()` pattern seems so fragile to me.

Reply:
> We should add a comment here that (hopefully) scope.close() will sync object state to make this remove effective.

I moved all `scope.Close()` to their respective functions in the reconciler.